### PR TITLE
add edge cases; fix soul_transfer bugs

### PR DESCRIPTION
**Pull request by sczembor**
- add unit tests to check the events and the edge cases
- update the comments 

Bugs:
- wrong index in the `last` transferred token
- we were using `token_id` in place of `class_id`

The limit for `_soul_transfer` that does not exceeds the max_gas_usage per call is `25`

### CODE REVIEW

File: contracts/registry/src/lib.rs [hunk 1]
```diff
@@ -170,12 +171,12 @@ impl Contract {
...
             }
         } else {
-            let last = &batch[token_counter];
+            let last = &batch[token_counter - 1];
             self.ongoing_soul_tx.insert(
                 &owner,
                 &IssuerTokenId {
                     issuer_id: last.0.issuer_id,
-                    token: last.1,
+                    token: last.0.class_id,
```
Comment (robert-zaremba) [inline contracts/registry/src/lib.rs:179]:
```suggestion
                    token: last.0.class_id,  // we reise IssuerTokenId type here (to not generate new code), but we store class_id instead of token here.
```

File: contracts/registry/src/lib.rs [hunk 2]
```diff
@@ -723,12 +742,175 @@ mod tests {
...
+
+        ctx.prepaid_gas = max_gas;
+        testing_env!(ctx.clone());
+        ctr.sbt_mint(vec![(dan(), batch_metadata[..10].to_vec())]);
+        assert_eq!(ctr.sbt_supply_by_owner(dan(), issuer1(), None), 10);
+
+        // soul transfer alice->alice2
+        ctx.predecessor_account_id = alice();
+
+        let limit: u32 = 25; //anything above this limit will fail due to exceeding maximum gas usage per call
```
Comment (robert-zaremba) [inline contracts/registry/src/lib.rs:843]: let's add a test to confirm that trying to transfer 30 tokens will fail with gas by using `should_panic`.
example:
```
    #[should_panic(expected = "less than or equal to 100")]
```
https://doc.rust-lang.org/book/ch11-01-writing-tests.html#checking-for-panics-with-should_panic

File: contracts/registry/src/lib.rs [hunk 3]
```diff
@@ -723,12 +742,175 @@ mod tests {
...
+        loop {
+            ctx.prepaid_gas = max_gas;
+            testing_env!(ctx.clone());
+            let result = ctr._sbt_soul_transfer(alice2(), limit as usize);
+            if result.1 == true {
+                break;
+            }
+        }
```
Comment (robert-zaremba) [inline contracts/registry/src/lib.rs:851]:
```suggestion
        while result.1 {
            ctx.prepaid_gas = max_gas;
            testing_env!(ctx.clone());
            let result = ctr._sbt_soul_transfer(alice2(), limit as usize);
        }
```

File: contracts/registry/src/lib.rs [hunk 4]
```diff
@@ -723,12 +742,175 @@ mod tests {
...
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer1(), None), 70);
+        assert_eq!(ctr.sbt_supply_by_owner(alice2(), issuer2(), None), 50);
+        assert_eq!(ctr.sbt_supply_by_owner(bob(), issuer1(), None), 20);
+        assert_eq!(ctr.sbt_supply_by_owner(carol(), issuer1(), None), 20);
+        assert_eq!(ctr.sbt_supply_by_owner(dan(), issuer1(), None), 10);
+    }
+
+    #[test]
+    fn soul_transfer_limit_basics() {
+        let max_gas: Gas = Gas::ONE_TERA.mul(300);
```
Comment (robert-zaremba) [inline contracts/registry/src/lib.rs:865]: seams we are reusing this, let's declare it in the `mod test { ...`

File: contracts/registry/src/lib.rs (hunk 2 above)
Comment (sczembor) [reply contracts/registry/src/lib.rs:847]: DONE

File: contracts/registry/src/lib.rs (hunk 4 above)
Comment (sczembor) [reply contracts/registry/src/lib.rs]: DONE
